src/renderer/metal.rs

```diff
@@ -0,0 +1,208 @@
+        surfaces::wrap_backend_render_target,
+        DirectContext, SurfaceOrigin,
+    },
+    Canvas, ColorSpace, ColorType, Surface,
+};
+use winit::{event_loop::EventLoopProxy, window::Window};
+
+use crate::{
+    profiling::tracy_gpu_zone,
+    renderer::{SkiaRenderer, VSync},
+    window::{macos::get_ns_window, UserEvent},
+};
+
+struct MetalDrawableSurface {
+    pub _drawable: Retained<ProtocolObject<dyn CAMetalDrawable>>,
+    pub metal_drawable: Retained<ProtocolObject<dyn MTLDrawable>>,
+    pub surface: Surface,
+}
+
+impl MetalDrawableSurface {
+    fn new(
+        drawable: Retained<ProtocolObject<dyn CAMetalDrawable>>,
+        context: &mut DirectContext,
+    ) -> MetalDrawableSurface {
+        tracy_gpu_zone!("MetalDrawableSurface.new");
+
+        let texture = unsafe { drawable.texture() };
+        let texture_info = unsafe { TextureInfo::new(Retained::as_ptr(&texture).cast()) };
+        let backend_render_target = gpu::backend_render_targets::make_mtl(
+            (texture.width() as i32, texture.height() as i32),
+            &texture_info,
+        );
+
+        let metal_drawable =
+            unsafe { Retained::cast::<ProtocolObject<dyn MTLDrawable>>(drawable.clone()) };
+
+        let surface = wrap_backend_render_target(
+            context,
+            &backend_render_target,
+            SurfaceOrigin::TopLeft,
+            ColorType::BGRA8888,
+            ColorSpace::new_srgb(),
+            None,
+        )
+        .expect("Failed to create skia surface with metal drawable.");
+
+        MetalDrawableSurface {
+            _drawable: drawable,
+            metal_drawable,
+            surface,
+        }
+    }
+
+    fn metal_drawable(&self) -> &ProtocolObject<dyn MTLDrawable> {
+        &self.metal_drawable
+    }
+}
+
+pub struct MetalSkiaRenderer {
+    window: Window,
+    _device: Retained<ProtocolObject<dyn MTLDevice>>,
+    command_queue: Retained<ProtocolObject<dyn MTLCommandQueue>>,
+    metal_layer: Retained<CAMetalLayer>,
+    _backend: BackendContext,
+    context: DirectContext,
+    metal_drawable_surface: Option<MetalDrawableSurface>,
+}
+
+impl MetalSkiaRenderer {
+    pub fn new(window: Window, srgb: bool, vsync: bool) -> Self {
+        log::info!("Initialize MetalSkiaRenderer...");
+
+        let draw_size = window.inner_size();
+        let ns_window = get_ns_window(&window);
+
+        let device = unsafe {
+            Retained::retain(MTLCreateSystemDefaultDevice())
+                .expect("Failed to create Metal system default device.")
+        };
+        let metal_layer = unsafe {
+            let metal_layer = CAMetalLayer::new();
+            metal_layer.setDevice(Some(&device));
+            metal_layer.setPixelFormat(if srgb {
+                MTLPixelFormat::BGRA8Unorm_sRGB
+            } else {
+                MTLPixelFormat::BGRA8Unorm
+            });
+            metal_layer.setPresentsWithTransaction(false);
+            metal_layer.setFramebufferOnly(false);
```

Comment:
I wonder if setting the `setFramebufferOnly(true)` would improve performance?

Reply:
I just simply follow `skia-safe` example at https://github.com/rust-skia/rust-skia/blob/f5d0181db8f7383539553f3daa4064c8f85cb993/skia-safe/examples/metal-window/main.rs#L158. I have no idea about this. But if you can prove this, we can change it to true definitely.

Reply:
And this is the comment in the example on this. I copy it here for convenience.

```rust
// Disabling this option allows Skia's Blend Mode to work.
// More about: https://developer.apple.com/documentation/quartzcore/cametallayer/1478168-framebufferonly
layer.set_framebuffer_only(false);
```

Reply:
OK, I see. Set it to `false` (the default value) improves performance. The disadvantage is

> However, you may not sample, read from, or write to those textures.

So if we don't use such functions, then we can remove it and leave it `false` as default.

Reply:
we can leave it as false then, my question was also out of curiosity :)

Reply:
I think skia internally need to read from these when you show a semi-transparent floating window for example, so it needs to be set to false.

Reply:
yes @fredizzimo is right, it definitely need to be false, it breaks the floating window blur. just checked myself.